### ISSUE TITLE
Spec: Multiple simultaneous NSM Forwarders

### ISSUE THREAD

**Issue by edwarnicke** [networkservicemesh/networkservicemesh#1565]
## Overview

Network Service Mesh on K8s delegates from the Network Service Manager (NSMgr) to the  NSM Forwarder (formerly NSM Dataplane) performing the cross connect between the mechanisms from each side of a vWire.  Examples:
* (kernel interface, memif)
* (kernel interface, vxlan)
* (vxlan, memif)
* (SRv6, kernel interface)
* (SRIOV VF, vlan)
* (HW NIC, vxlan)
The NSM Forwarder is the only portion of the NSM system in K8s that has privilege.

Network Service Mesh was always intended to support multiple simultaneous NSM Forwarders 

Supporting multiple NSM Forwarders is crucial to properly supporting a broad array of both local and remote mechanisms.  Requiring all mechanisms to be implemented by a single monolithic NSM Forwarder presents a major bottleneck in the system.  This is exacerbated by the expectation that for the SRIOV VF case, the NSM Forwarder may be particular to the NIC itself, not SRIOV generically (due to complexities of programming the NIC to, for example, always encap decap a particular vlan to/from a particular VF).

Currently, we do Mechanism discovery via the [MonitorMechanisms](https://github.com/networkservicemesh/networkservicemesh/blob/master/dataplane/pkg/apis/dataplane/dataplane.proto#L22) API call from the NSMgr to the NSM Forwarder.

The problem with MonitorMechanisms is that:

1. It is not guaranteed that because a NSM Forwarder supports Mechanism1 and Mechanism2 that it supports the cross connect between Mechanism1 and Mechanism2
2. The NSM Forwarder may need to only provide support for a particular MechanismType when certain parameters are met.  Example: an SRIOV VF NSM Forwarder that only supports a subset of VF devices... because its NIC specific and the Node has multiple NICs (with a different NSM Forwarder per NIC).

To fix this, remove MonitorMechanisms, and instead introduce:

```proto
service Dataplane {
  rpc Request(crossconnect.CrossConnect) returns (crossconnect.CrossConnect);
  rpc Close(crossconnect.CrossConnect) returns (google.protobuf.Empty);
  rpc Available(repeated crossconnect.CrossConnect) returns (repeated crossconnect.CrossConnect)
}
```

The flow becomes:
![MultiForwarder Support diagrams (1)](https://user-images.githubusercontent.com/1455208/64261330-f6871980-cef1-11e9-9ff5-70b9df0e8797.png)

Basically... whenever a Request comes in (either local or remote), the NSMgr ask the NSMForwarders (in parallel) what the available CrossConnects are.  

1. The crossconnect.CrossConnect passed to Available should not contain a dst Mechanism, and if one is set it will be ignored by the NSM Forwarder.
2. Each crossconnect.CrossConnect returned by the NSM Forwarder will have a dst Mechanism set.
3.  NSMgr is still authoritative as to crossconnect ids.

Once the NSMgr has the return from Available on each NSM Forwarder, it can synthesize a Mechanism list for its next Request (whether to a peer NSMgr or an NSE).

**Comment by edwarnicke** [comment]
@rdimitrov @nickolaev Could you guys have a look at this?

**Comment by rdimitrov** [comment]
I like the idea of polling for available CrossConnects 👍 What comes to my mind is:

1. How do we choose a forwarding plane in case two or more forwading planes return the same availabilities? For ex, kernel interface as a destination mechanism.

2. In case we have remote source <-> local destination. How do we choose a forwarding plane if one supports "vxlan <-> memif" and the other one has "vxlan <-> kernel interface"?

**Comment by edwarnicke** [comment]
@rdimitrov #2 I think is fairly easy.  If you have FW1 (NSM Forwarder 1) and FW2 (NSM Fowarder2) and:

FW1 -> Avaliable(VXLAN) -> [memif]
FW2 -> Available(VXLAN) -> [kernel]

Then both get sent as options to the NSE (note: there's an order of preference question here).  The NSE then selects whether it prefers memif or kernel.  If memif, the NSMgr can only get that with FW1, so it picks that (or vis-versa if kernel).

So I think the issue really comes down to two kinds of preference order:

1.  Your case #1: Given two Forwarders that have the same ability to provide a selected CrossConnection... which do you pick and how?
2.  If you have mechanisms available from two different Forwarders, how do you order them in the Mechanism preference list to the NSE (or if remote, peer NSMgr)?

I can think of a couple of ways to handle these, neither of which I much like, so I'm very interested in other ideas:

1.  Let the Forwarders assert a priority... 
2. Let the NSMgr specify a priority
3. Synthesize a priority by something arbitrary (like order registered)

@rdimitrov Do you have any other ideas here... I'm in the market for something clever :)

**Comment by rdimitrov** [comment]
We had a brief discussion with Nikolay last week and I think we were brainstorming mostly around the 1st option. Each mechanism for each forwarder comes with a priority and it's up to the NSMgr to take that into account while processing a request. Though that might result in some configuration mess, so I agree that this might not be the best way possible.

An assumption I have is that the administrator deploying NSM with all the forwarding planes should be the one setting up the priorities, right? Or that's not always the case?

**Comment by edwarnicke** [comment]
@rdimitrov How about this... we define a *default* priority that NSM Forwarders are expected to use unless *configured* to use a different one.  So say default priority is 50.  Every NSM Forwarder comes in with priority 50... unless configured to a higher or lower priority.

The question then is, how do we break the tie for default?

**Comment by rdimitrov** [comment]
@edwarnicke In that sense, which would be the "unit" that we want to prioritize - forwarding plane, connection type or mechanism type?

**Comment by edwarnicke** [comment]
> @edwarnicke In that sense, which would be the "unit" that we want to prioritize - forwarding plane, connection type or mechanism type?

@rdimitrov
